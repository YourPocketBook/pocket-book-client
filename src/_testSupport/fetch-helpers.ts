import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
