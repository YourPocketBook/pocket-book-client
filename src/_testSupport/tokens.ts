export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
