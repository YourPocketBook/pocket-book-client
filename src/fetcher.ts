import isFuture from "date-fns/isFuture";
import jwtDecode from "jwt-decode";

class Fetcher {
  private realName: string = "";
  private nameChanged?: (name: string) => void;
  private token: string | null;
  private parsedToken:
    | { exp: number; given_name: string; role: string }
    | undefined;

  constructor() {
    this.token = localStorage.getItem("token");
    this.parseToken();
  }

  public setNameChangeHandler(func?: (name: string) => void) {
    this.nameChanged = func;
  }

  public authFetch(input: RequestInfo, init?: RequestInit): Promise<Response> {
    if (!this.token) {
      throw new Error("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${this.token}`
      }
    };

    return fetch(input, init);
  }

  public saveToken(token: string) {
    localStorage.setItem("token", token);
    this.token = token;
    this.parseToken();
  }

  public checkToken(token: string): boolean {
    return this.token === token;
  }

  public clearToken() {
    localStorage.removeItem("token");
    this.token = null;
    this.parsedToken = undefined;
  }

  public getName() {
    return this.realName;
  }

  public updateName(name: string) {
    this.realName = name;
    if (this.nameChanged) {
      this.nameChanged(name);
    }
  }

  public isAdmin() {
    if (this.parsedToken) {
      return this.parsedToken.role === "Admin";
    }
    return false;
  }

  public hasToken(): boolean {
    return !!this.token;
  }

  public isInDate() {
    if (this.parsedToken) {
      return isFuture(this.parsedToken.exp * 1000);
    }

    return false;
  }

  private parseToken() {
    if (this.token) {
      try {
        this.parsedToken = jwtDecode(this.token);
      } catch {
        this.token = null;
        this.updateName("");
        localStorage.removeItem("token");
        this.parsedToken = undefined;
      }

      if (this.parsedToken) {
        this.updateName(this.parsedToken.given_name);
      }
    } else {
      this.parsedToken = undefined;
      this.updateName("");
    }
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
