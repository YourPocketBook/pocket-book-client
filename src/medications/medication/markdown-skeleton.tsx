import React from "react";
import Skeleton from "react-loading-skeleton";
import ReactMarkdown from "react-markdown";

export function MarkdownSkeleton({ source }: { source?: string }) {
  return source ? <ReactMarkdown source={source} /> : <Skeleton count={3} />;
}
