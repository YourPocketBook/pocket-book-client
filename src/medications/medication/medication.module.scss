@import "../../styles/colors.scss";

.background {
  background-color: $color-sja-dark-green;
  min-height: 100vh;
  position: relative;
  padding: 10px;
}

.centreBox {
  background-color: $color-sja-white;
  width: calc(100%-20px);
  border-radius: 5px;
  padding: 10px;
}

.p {
  h1 {
    text-align: center;
    color: $color-sja-dark-green;
  }

  h2 {
    text-align: center;
    color: $color-sja-black;
    font-size: 1.5rem;
  }

  h3 {
    color: $color-sja-black;
    font-size: 1.5rem;
  }

  p,
  li {
    color: $color-sja-black;
    font-size: 1.25rem;
  }
}

footer {
  margin-bottom: 6px;
  color: lighten($color: $color-sja-black, $amount: 50);
}
