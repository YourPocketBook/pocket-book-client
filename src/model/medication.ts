export interface IMedication {
  id: number;
  name: string;
  indications: string;
  inclusionCriteria: string;
  exclusionCriteria: string;
  form: string;
  route: string;
  dose: string;
  sideEffects: string;
  adviceIfTaken: string;
  adviceIfDeclined: string;
  policyDate?: string | null;
}
