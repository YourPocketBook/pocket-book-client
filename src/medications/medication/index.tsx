import format from "date-fns/format";
import parseDate from "date-fns/parseISO";
import React, { useState } from "react";
import Button from "reactstrap/lib/Button";
import ButtonGroup from "reactstrap/lib/ButtonGroup";
import Col from "reactstrap/lib/Col";
import Modal from "reactstrap/lib/Modal";
import ModalBody from "reactstrap/lib/ModalBody";
import ModalFooter from "reactstrap/lib/ModalFooter";
import ModalHeader from "reactstrap/lib/ModalHeader";
import Row from "reactstrap/lib/Row";
import { fetcher } from "../../fetcher";
import { history } from "../../history";
import { useMedication } from "../../hooks/useMedications";
import { useOnlineStatus } from "../../hooks/useOnlineStatus";
import { Link } from "../../link";
import { PrivatePage } from "../../private-page";
import { MarkdownSkeleton } from "./markdown-skeleton";
import styles from "./medication.module.scss";

const deleteEndPoint = "/api/delete-medication";

/**
 * Component that displays the medication to the user and allows administrators
 * to edit or delete medications.
 *
 * Invalid or expired token is handled by [[PrivatePage]].
 *
 * If the user is offline, it displays the offline warning and disables
 * the edit and delete buttons.
 *
 * If load into state returns nothing or fails, it displays an error message.
 *
 * While the delete request is running, it disables the submit button
 * and displays a "busy" indicator.
 * When the delete request completes, it redirects to the medication list.
 */
export function Medication({ id, reload }: { id: string; reload: boolean }) {
  const { medication, isLoading, loadFailed } = useMedication(id, reload);
  const [isBusy, setIsBusy] = useState(false);

  let content;

  const isAdmin = fetcher.isAdmin();
  const isOnline = useOnlineStatus();
  const [showDeleteConfirm, setShowDeleteConfirm] = useState(false);

  function deleteMedication() {
    setIsBusy(true);

    fetcher
      .authFetch(`${deleteEndPoint}/${id}`, {
        method: "DELETE"
      })
      .then(() => {
        setIsBusy(false);
        history.push("/medications");
      });
  }

  if (isLoading && !medication) {
    content = <div>Loading...</div>;
  } else if (!isLoading && (loadFailed || !medication)) {
    content = (
      <React.Fragment>
        <h1>Error Loading Medication</h1>
        <p>We weren't able to load that medication. Try refreshing the page.</p>
      </React.Fragment>
    );
  } else if (medication) {
    content = (
      <React.Fragment>
        <Row>
          <Col>
            <Link
              className="btn btn-link text-decorate-underline"
              href="/medications"
            >
              Go back...
            </Link>
          </Col>
          <Col>
            <h1 className="text-capitalize">
              {(medication.name || "").toLowerCase()}
            </h1>
          </Col>
          <Col />
        </Row>
        <h2>Clinical Condition</h2>
        <h3>Indication</h3>
        <MarkdownSkeleton source={medication.indications} />
        <h3>Inclusion Criteria</h3>
        <MarkdownSkeleton source={medication.inclusionCriteria} />
        <h3>Exclusion Criteria</h3>
        <MarkdownSkeleton source={medication.exclusionCriteria} />
        <h2>Medicine Details</h2>
        <h3>Name, Form and Strength of Medicine</h3>
        <MarkdownSkeleton source={medication.form} />
        <h3>Route or Method</h3>
        <MarkdownSkeleton source={medication.route} />
        <h3>Dosage</h3>
        <MarkdownSkeleton source={medication.dose} />
        <h3>Side Effects</h3>
        <MarkdownSkeleton source={medication.sideEffects} />
        <h3>Advice to Patient and/or Carer</h3>
        <MarkdownSkeleton source={medication.adviceIfTaken} />
        <h3>Advice if Patient Declines or is Excluded</h3>
        <MarkdownSkeleton source={medication.adviceIfDeclined} />
        <h2>Audit Trail</h2>
        <h3>Records</h3>
        <p>
          All details should be recorded on the Patient Report Form and any
          adverse reactions reported via the Incident Management Framework.
        </p>
        <footer>
          {medication.policyDate
            ? `Taken from Medications Administration Directives dated ${format(
                parseDate(medication.policyDate),
                "d MMM yyyy"
              )}.`
            : null}
        </footer>
        {isAdmin ? (
          <ButtonGroup className="text-center">
            <Link
              className="btn btn-dark-green btn-lg"
              href={`/admin/edit-medication/${id}`}
              disabled={!isOnline}
            >
              Edit Medication
            </Link>
            <Button
              className="btn btn-danger btn-lg"
              onClick={() => setShowDeleteConfirm(true)}
              disabled={!isOnline}
            >
              Delete Medication
            </Button>
          </ButtonGroup>
        ) : null}
        <Modal
          isOpen={showDeleteConfirm}
          toggle={() => setShowDeleteConfirm(!showDeleteConfirm)}
        >
          <ModalHeader>Are you sure?</ModalHeader>
          <ModalBody>
            Are you sure you want to delete '{medication.name}'? This cannot be
            un-done.
          </ModalBody>
          <ModalFooter>
            <Button color="danger" onClick={deleteMedication} disabled={isBusy}>
              Yes
            </Button>
            <Button
              color="secondary"
              onClick={() => setShowDeleteConfirm(false)}
            >
              No
            </Button>
          </ModalFooter>
        </Modal>
      </React.Fragment>
    );
  }

  return (
    <PrivatePage>
      <div className={styles.background}>
        <div className={styles.centreBox}>
          <div className={styles.p}>{content}</div>
        </div>
      </div>
    </PrivatePage>
  );
}
