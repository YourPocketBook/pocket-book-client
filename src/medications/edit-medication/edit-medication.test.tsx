import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, wait } from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { EditMedication } from ".";
import {
  authenticatedGet,
  authenticatedPut
} from "../../_testSupport/fetch-helpers";
import {
  checkForClass,
  checkState,
  testDefaultState
} from "../../_testSupport/medication-form-helpers";
import { inDateAdminToken } from "../../_testSupport/tokens";
import { fetcher } from "../../fetcher";
import { history } from "../../history";
import { IMedication } from "../../model/medication";

const editEndpoint = "/api/update-medication";
const getEndpoint = "/api/get-medication";

describe("Edit Medication Page", () => {
  beforeEach(() => {
    fetcher.saveToken(inDateAdminToken);
    history.push = jest.fn();
  });

  afterEach(() => {
    fetcher.clearToken();
    localStorage.clear();
  });

  it("renders without crashing", () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    const div = document.createElement("div");
    ReactDOM.render(<EditMedication id="42" />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders loading correctly", () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateAdminToken,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { container } = render(<EditMedication id="42" />);

    expect(container).toMatchSnapshot();
  });

  it("renders loaded correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);

    const { container } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("defaults to correct values", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);

    const { getByLabelText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    checkState(getByLabelText, true);
  });

  it("declares valid values correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(`${editEndpoint}/42`, inDateAdminToken, testDefaultState, {
      body: JSON.stringify([{ path: "doses", error: "test" }]),
      status: 400
    });

    const { getByLabelText, getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    checkForClass(getByLabelText, "is-valid");
  });

  it("declares server side invalid values correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(`${editEndpoint}/42`, inDateAdminToken, testDefaultState, {
      body: JSON.stringify([
        { path: "adviceIfDeclined", error: "isInvalid" },
        { path: "adviceIfTaken", error: "isBlank" },
        { path: "dose", error: "isInvalid" },
        { path: "exclusionCriteria", error: "isBlank" },
        { path: "form", error: "isInvalid" },
        { path: "inclusionCriteria", error: "isBlank" },
        { path: "indications", error: "isInvalid" },
        { path: "name", error: "isBlank" },
        { path: "policyDate", error: "isBlank" },
        { path: "route", error: "isInvalid" },
        { path: "sideEffects", error: "isBlank" }
      ]),
      status: 400
    });

    const { getByLabelText, getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    checkForClass(getByLabelText, "is-invalid");
  });

  it("declares invalid values correctly", async () => {
    const testInvalidState = {
      adviceIfDeclined: "",
      adviceIfTaken: "",
      dose: "",
      exclusionCriteria: "",
      form: "",
      inclusionCriteria: "",
      indications: "",
      name: "",
      policyDate: "",
      route: "",
      sideEffects: ""
    };

    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testInvalidState);

    const { getByLabelText, getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    fireEvent.click(getByText("Save"));

    checkForClass(getByLabelText, "is-invalid");
  });

  it("renders server error correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(`${editEndpoint}/42`, inDateAdminToken, testDefaultState, {
      status: 500
    });

    const { getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/The server has had an issue/)).toBeInTheDocument();
  });

  it("renders in progress correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(
      `${editEndpoint}/42`,
      inDateAdminToken,
      testDefaultState,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    expect(getByText("Save")).toBeDisabled();
  });

  it("reports name already in use error", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(`${editEndpoint}/42`, inDateAdminToken, testDefaultState, {
      body: JSON.stringify([{ path: "name", error: "isInUse" }]),
      status: 400
    });

    const { getByLabelText, getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("Name")).toHaveClass("is-invalid");
  });

  it("redirects on successful form submit", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedPut(`${editEndpoint}/42`, inDateAdminToken, testDefaultState, {
      id: 10
    });

    const { getByText } = render(<EditMedication id="42" />);

    await wait(() => expect(fetchMock.done(`${getEndpoint}/42`)).toBeTruthy());

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(history.push).toBeCalledWith("/medication/10?reload=true");

    const cachedString = localStorage.getItem("medications");

    expect(cachedString).not.toBeNull();

    const cachedData = JSON.parse(cachedString as string) as Array<
      Partial<IMedication>
    >;

    const cachedItem = cachedData.find(m => m.id === 10);

    expect(cachedItem).toBeDefined();
    expect(cachedItem).toEqual({ id: 10, ...testDefaultState });
  });
});
