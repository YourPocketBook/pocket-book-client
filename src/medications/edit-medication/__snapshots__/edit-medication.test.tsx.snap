// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Medication Page renders loaded correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="formBackground container"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <h1
              class="text-center"
            >
              Edit Medication
            </h1>
          </div>
        </div>
        <form
          novalidate=""
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="name"
                >
                  Name
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="name"
                  name="name"
                  placeholder="Name"
                  type="text"
                  value="Test Name"
                />
                <div
                  class="invalid-feedback"
                >
                  You need to enter the name of the medication.
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="policyDate"
                >
                  Policy Date
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="policyDate"
                  name="policyDate"
                  placeholder="Policy Date"
                  type="date"
                  value="2019-12-10"
                />
                <div
                  class="invalid-feedback"
                >
                  You need to enter the date of the policy this is taken from.
                </div>
              </div>
            </div>
          </div>
          <h2
            class="h5"
          >
            Clinical Condition
          </h2>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="indications"
                >
                  Indications
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="indications"
                  name="indications"
                  placeholder="Indications"
                >
                  Test Indications
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the indications for the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Indications
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="inclusionCriteria"
                >
                  Inclusion Criteria
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="inclusionCriteria"
                  name="inclusionCriteria"
                  placeholder="Inclusion Criteria"
                >
                  Test Inclusion
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the inclusion criteria for the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Inclusion
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="exclusionCriteria"
                >
                  Exclusion Criteria
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="exclusionCriteria"
                  name="exclusionCriteria"
                  placeholder="Exclusion Criteria"
                >
                  Test Exclusion
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the exclusion criteria for the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Exclusion
              </p>
            </div>
          </div>
          <h2
            class="h5"
          >
            Medicine Details
          </h2>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="form"
                >
                  Name, Forms and Strengths of Medicine
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="form"
                  name="form"
                  placeholder="Name, Forms and Strengths of Medicine"
                >
                  Test Form
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the name, forms and strengths of the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Form
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="route"
                >
                  Route or Method
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="route"
                  name="route"
                  placeholder="Route or Method"
                >
                  Test Route
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the route or method for the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Route
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="dose"
                >
                  Dosage
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="dose"
                  name="dose"
                  placeholder="Dosage"
                >
                  Test Dose
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the dose for the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Dose
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="sideEffects"
                >
                  Side Effects
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="sideEffects"
                  name="sideEffects"
                  placeholder="Side Effects"
                >
                  Test Side Effects
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the side effects of the medication.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Side Effects
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="adviceIfTaken"
                >
                  Advice if Taken
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="adviceIfTaken"
                  name="adviceIfTaken"
                  placeholder="Advice if Taken"
                >
                  Test Advice 2
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the advice to give if the medication is taken.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Advice 2
              </p>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="adviceIfDeclined"
                >
                  Advice if Declined
                </label>
                <textarea
                  autocomplete="off"
                  class="h-100 form-control"
                  id="adviceIfDeclined"
                  name="adviceIfDeclined"
                  placeholder="Advice if Declined"
                >
                  Test Advice 1
                </textarea>
                <div
                  class="invalid-feedback"
                >
                  You need to enter the advice to give if the medication is declined.
                </div>
              </div>
            </div>
            <div
              class="col-md"
            >
              <p>
                Test Advice 1
              </p>
            </div>
          </div>
          <button
            class="btn btn-dark-green btn-block"
            type="submit"
          >
            Save
          </button>
        </form>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Edit Medication Page renders loading correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="formBackground container"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <h1
              class="text-center"
            >
              Edit Medication
            </h1>
          </div>
        </div>
        <div>
          One moment please...
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
